From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: kylechar <kylechar@chromium.org>
Date: Tue, 8 Sep 2026 11:01:10 -0700
Subject: Check transfer cache entries for wrap around

The GPU service transfer cache uses the same namespace of ids for every
client in the same process, The ID generation had no protection against
wrap around so we could see duplicate IDs if enough are allocated
maliciously. To prevent ID collisions add a CHECK that ensures wrap
around doesn't happen. We don't expect wrap around to happen through
normal usage of Chrome.

This CL moves the ID generation implementation to source file since it's
only used there. It also generates uint32_t instead of int values so the
full uint32_t ID range is used.

Bug: 556260782
Change-Id: I876c5fdd55185a3a2d290abfdb59cb257f8e1196
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/8349499
Commit-Queue: Kyle Charbonneau <kylechar@chromium.org>
Reviewed-by: Vladimir Levin <vmpstr@chromium.org>
Cr-Commit-Position: refs/heads/main@{#1693922}

diff --git a/cc/paint/image_transfer_cache_entry.cc b/cc/paint/image_transfer_cache_entry.cc
index 729563a59c023c2edba5da410e2bd05d806d4581..648e6240189668074927415fd9c3c36a49367690 100644
--- a/cc/paint/image_transfer_cache_entry.cc
+++ b/cc/paint/image_transfer_cache_entry.cc
@@ -6,9 +6,11 @@
 
 #include <algorithm>
 #include <array>
+#include <limits>
 #include <type_traits>
 #include <utility>
 
+#include "base/atomic_sequence_num.h"
 #include "base/compiler_specific.h"
 #include "base/containers/span.h"
 #include "base/functional/callback_helpers.h"
@@ -434,6 +436,13 @@ sk_sp<SkImage> ReadImage(
   }
 }
 
+uint32_t GetNextCacheEntryId() {
+  static base::AtomicSequenceNumberT<uint32_t> id_sequence;
+  uint32_t id = id_sequence.GetNext();
+  CHECK_NE(id, std::numeric_limits<uint32_t>::max());
+  return id;
+}
+
 }  // namespace
 
 size_t NumberOfPlanesForYUVDecodeFormat(YUVDecodeFormat format) {
@@ -496,7 +505,7 @@ ClientImageTransferCacheEntry::ClientImageTransferCacheEntry(
     sk_sp<SkColorSpace> target_color_space)
     : needs_mips_(needs_mips),
       target_color_space_(target_color_space),
-      id_(GetNextId()),
+      id_(GetNextCacheEntryId()),
       image_(image) {
   ComputeSize();
 }
@@ -507,7 +516,7 @@ ClientImageTransferCacheEntry::ClientImageTransferCacheEntry(
     const SkGainmapInfo& gainmap_info,
     bool needs_mips)
     : needs_mips_(needs_mips),
-      id_(GetNextId()),
+      id_(GetNextCacheEntryId()),
       image_(image),
       gainmap_image_(gainmap_image),
       gainmap_info_(gainmap_info) {
@@ -516,9 +525,6 @@ ClientImageTransferCacheEntry::ClientImageTransferCacheEntry(
 
 ClientImageTransferCacheEntry::~ClientImageTransferCacheEntry() = default;
 
-// static
-base::AtomicSequenceNumber ClientImageTransferCacheEntry::s_next_id_;
-
 uint32_t ClientImageTransferCacheEntry::SerializedSize() const {
   return size_;
 }
diff --git a/cc/paint/image_transfer_cache_entry.h b/cc/paint/image_transfer_cache_entry.h
index 3e00775bc8d51da494d3a874c880ed7e234c691a..14a8efdda0aedd8ae88284fda2520506e1b189fb 100644
--- a/cc/paint/image_transfer_cache_entry.h
+++ b/cc/paint/image_transfer_cache_entry.h
@@ -12,7 +12,6 @@
 #include <optional>
 #include <vector>
 
-#include "base/atomic_sequence_num.h"
 #include "base/containers/span.h"
 #include "base/memory/raw_ptr.h"
 #include "cc/paint/tone_map_util.h"
@@ -96,7 +95,6 @@ class CC_PAINT_EXPORT ClientImageTransferCacheEntry final
   uint32_t SerializedSize() const final;
   bool Serialize(base::span<uint8_t> data) const final;
 
-  static uint32_t GetNextId() { return s_next_id_.GetNext(); }
   bool IsYuv() const {
     return image_.yuv_plane_config != SkYUVAInfo::PlaneConfig::kUnknown;
   }
@@ -109,7 +107,6 @@ class CC_PAINT_EXPORT ClientImageTransferCacheEntry final
   sk_sp<SkColorSpace> target_color_space_;
   const uint32_t id_;
   uint32_t size_ = 0;
-  static base::AtomicSequenceNumber s_next_id_;
 
   Image image_;
 
diff --git a/gpu/command_buffer/service/shared_context_state.cc b/gpu/command_buffer/service/shared_context_state.cc
index d644576c364b8bb6a026aa8c9d35823489a7b956..c723443f8cac54d710557754cbdcf4b4556609f5 100644
--- a/gpu/command_buffer/service/shared_context_state.cc
+++ b/gpu/command_buffer/service/shared_context_state.cc
@@ -4,6 +4,7 @@
 
 #include "gpu/command_buffer/service/shared_context_state.h"
 
+#include "base/atomic_sequence_num.h"
 #include "base/compiler_specific.h"
 #include "base/debug/dump_without_crashing.h"
 #include "base/immediate_crash.h"
